java/runtime/src/main/java/io/ray/runtime/io/MemoryBuffer.java

```diff
@@ -0,0 +1,1018 @@
+   * The buffer must be at least of size 8 bytes.
+   *
+   * @param buffer The byte array whose memory is represented by this memory buffer.
+   * @param offset The offset of the sub array to be used; must be non-negative and no larger than
+   *     <tt>array.length</tt>.
+   * @param length buffer size
+   */
+  MemoryBuffer(byte[] buffer, int offset, int length) {
+    Preconditions.checkArgument(offset >= 0 && length >= 0);
+    if (offset + length > buffer.length) {
+      throw new IllegalArgumentException(
+          String.format("%d exceeds buffer size %d", offset + length, buffer.length));
+    }
+    initHeapBuffer(buffer, offset, length);
+  }
+
+  /**
+   * Creates a new memory buffer that represents the memory of the byte array.
+   *
+   * <p>The memory buffer references the given owner.
+   *
+   * @param buffer The byte array whose memory is represented by this memory buffer.
+   */
+  MemoryBuffer(byte[] buffer) {
+    this(buffer, 0, buffer.length);
+  }
+
+  private void initHeapBuffer(byte[] buffer, int offset, int length) {
+    if (buffer == null) {
+      throw new NullPointerException("buffer");
+    }
+    this.heapMemory = buffer;
+    this.address = BYTE_ARRAY_BASE_OFFSET + offset;
+    this.size = length;
+    this.addressLimit = this.address + this.size;
+  }
+
+  /**
+   * Creates a new memory buffer that represents the memory backing the given direct byte buffer
+   * section of [buffer.position(), buffer,limit()). Note that the given ByteBuffer must be direct
+   * {@link ByteBuffer#allocateDirect(int)}, otherwise this method with throw an
+   * IllegalArgumentException.
+   *
+   * @param buffer The byte buffer whose memory is represented by this memory buffer.
+   * @throws IllegalArgumentException Thrown, if the given ByteBuffer is not direct.
+   */
+  MemoryBuffer(ByteBuffer buffer) {
+    this(Platform.getAddress(buffer) + buffer.position(), buffer.remaining());
+    this.offHeapBuffer = buffer;
+  }
+
+  /**
+   * Creates a new memory buffer that represents the memory at the absolute address given by the
+   * pointer.
+   *
+   * @param offHeapAddress The address of the memory represented by this memory buffer.
+   * @param size The size of this memory buffer.
+   */
+  MemoryBuffer(long offHeapAddress, int size) {
+    if (offHeapAddress <= 0) {
+      throw new IllegalArgumentException("negative pointer or size");
+    }
+    if (offHeapAddress >= Long.MAX_VALUE - Integer.MAX_VALUE) {
+      // this is necessary to make sure the collapsed checks are safe against numeric overflows
+      throw new IllegalArgumentException(
+          "Buffer initialized with too large address: "
+              + offHeapAddress
+              + " ; Max allowed address is "
+              + (Long.MAX_VALUE - Integer.MAX_VALUE - 1));
+    }
+
+    this.heapMemory = null;
+    this.address = offHeapAddress;
+    this.addressLimit = this.address + size;
+    this.size = size;
+  }
```

Comment:
We can move those static methods in MemoryUtils to this class, as factory methods. And simplify those constructors. For example, only keep 2 constructors, one for heap memory and the other for off heap memory.

Reply:
Good suggestions